bftengine/src/communication/TlsTCPCommunication.cpp

```diff
@@ -1044,19 +1220,23 @@ class TlsTCPCommunication::TlsTcpImpl {
   * On success, returns 0.
   */
   int Stop() {
-    if (!_pIoThread)
+    if (!_pIoThread) {
       return 0; // stopped
+    }
 
     _service.stop();
     _pIoThread->join();
-    _service.reset();
+
+    _connections.clear();
```

Comment:
For my understanding, calling _service.Stop() aborts all operations, prevents ready handlers to run - and the allows the AsyncTcpConnection object to be destroyed (using shared_from_this) 
after they are removed from the list.
We dont clean sockets explicitly, relying on Asio cleanup (both TLS stream and TCP)

From Asio docs:
" To effect a shutdown, the application will then need to call the io_service object's stop() member function. This will cause the io_service run() call to return as soon as possible, abandoning unfinished operations and without permitting ready handlers to be dispatched."